File: fm-dddrive/src/main/java/io/dddrive/core/property/model/impl/ReferenceSetPropertyImpl.java
package io.dddrive.core.property.model.impl;

import static io.dddrive.util.Invariant.assertThis;
import static io.dddrive.util.Invariant.requireThis;

import java.util.HashSet;
import java.util.Set;

import io.dddrive.core.ddd.model.Aggregate;
import io.dddrive.core.property.model.AggregateResolver;
import io.dddrive.core.property.model.EntityWithProperties;
import io.dddrive.core.property.model.EntityWithPropertiesSPI;
import io.dddrive.core.property.model.ReferenceSetProperty;
import io.dddrive.core.property.model.base.PropertyBase;

public class ReferenceSetPropertyImpl<A extends Aggregate> extends PropertyBase<A> implements ReferenceSetProperty<A> {

	private final Set<Object> itemSet = new HashSet<>();

	public ReferenceSetPropertyImpl(EntityWithProperties entity, String name, AggregateResolver<A> repository) {
		super(entity, name);
	}

	@Override
	public void clearItems() {
		requireThis(this.isWritable(), "not frozen");
		this.itemSet.forEach(this::removeItem);
		this.itemSet.clear();
		((EntityWithPropertiesSPI) this.getEntity()).doAfterClear(this);
	}

	@Override
	public void addItem(Object id) {
		requireThis(this.isWritable(), "not frozen");
		requireThis(id != null, "aggregateId not null");
		if (this.hasItem(id)) {
			return;
		}
		assertThis(this.isValidAggregateId(id), "valid aggregate id [" + id + "]");
		if (!this.hasItem(id)) {
			EntityWithPropertiesSPI entity = (EntityWithPropertiesSPI) this.getEntity();
			entity.fireValueAddedChange(this, id);
			this.itemSet.add(id);
			entity.doAfterAdd(this, null);
		}
	}

	@Override
	public Set<Object> getItems() {
		return Set.copyOf(this.itemSet);
	}

	@Override
	public boolean hasItem(Object aggregateId) {
		return this.itemSet.contains(aggregateId);
	}

	@Override
	public void removeItem(Object aggregateId) {
		requireThis(this.isWritable(), "not frozen");
		requireThis(aggregateId != null, "aggregateId not null");
		if (this.hasItem(aggregateId)) {
			EntityWithPropertiesSPI entity = (EntityWithPropertiesSPI) this.getEntity();
			entity.fireValueRemovedChange(this, aggregateId);
			this.itemSet.remove(aggregateId);
			entity.doAfterRemove(this);
		}
	}

	// TODO too expensive?
	private boolean isValidAggregateId(Object id) {
		return true;
	}

}

File: fm-dddrive/src/main/java/io/dddrive/core/property/model/impl/ReferenceSetPropertyImpl.kt
package io.dddrive.core.property.model.impl

import io.dddrive.core.ddd.model.Aggregate
import io.dddrive.core.ddd.model.AggregateRepository
import io.dddrive.core.ddd.model.AggregateRepositorySPI
import io.dddrive.core.property.model.EntityWithProperties
import io.dddrive.core.property.model.EntityWithPropertiesSPI
import io.dddrive.core.property.model.ReferenceSetProperty
import io.dddrive.core.property.model.base.PropertyBase

class ReferenceSetPropertyImpl<A : Aggregate>(
	entity: EntityWithProperties,
	name: String,
	val repo: AggregateRepository<A>,
) : PropertyBase<A>(entity, name),
	ReferenceSetProperty<A> {

	private val itemSet: MutableSet<Any> = mutableSetOf()

	override fun clearItems() {
		require(isWritable) { "writable" }
		items.forEach { removeItem(it) }
		check(itemSet.isEmpty()) { "all items removed" }
		itemSet.clear()
		(entity as EntityWithPropertiesSPI).doAfterClear(this)
	}

	override fun addItem(aggregateId: Any) {
		require(isWritable) { "writable" }
		require(isValidAggregateId(aggregateId)) { "valid aggregate id [$aggregateId]" }
		if (hasItem(aggregateId)) {
			return
		}
		if (!hasItem(aggregateId)) {
			val entity = entity as EntityWithPropertiesSPI
			entity.fireValueAddedChange(this, aggregateId)
			itemSet.add(aggregateId)
			entity.doAfterAdd(this, null)
		}
	}

	override val items: Set<Any>
		get() = itemSet.toSet()

	override fun hasItem(aggregateId: Any): Boolean = itemSet.contains(aggregateId)

	override fun removeItem(aggregateId: Any) {
		require(isWritable) { "writable" }
		require(isValidAggregateId(aggregateId)) { "valid aggregate id [$aggregateId]" }
		if (hasItem(aggregateId)) {
			val entity = entity as EntityWithPropertiesSPI
			entity.fireValueRemovedChange(this, aggregateId)
			itemSet.remove(aggregateId)
			entity.doAfterRemove(this)
		}
	}

	// TODO too expensive?
	private fun isValidAggregateId(id: Any): Boolean = (repo as AggregateRepositorySPI<*>).persistenceProvider.isValidId(id)

}
